Externalize the settings

0livd opened the pull request:
Default settings from app root path are loaded first
Settings are then overriden by /etc/ihatemoney/ihatemoney.cfg
or by another file which path is set in an env var
Fixes #187 

Readme should be updated if this gets merged

0livd commented:
yup it's nice to see that there is such a dynamic around a little project like this :)

Yes it's correct, maybe we can ditch the environment var thing and just try to load /etc/ihatemoney/ihatemoney.cfg but in this case I think we should create this directory by default on a production installation

almet commented:
I actually quite like the current approach with the env settings, as it enables some customisation for the users :-) My question was more about testing here :-)

0livd commented:
Oh ok I thought you said having multiple ways to load settings was too complicated :)

I'll try to get some inspiration from the flask test suite to write some tests
https://github.com/pallets/flask/blob/master/tests/test_config.py

JocelynDelalande commented:
> Yes it's correct, maybe we can ditch the environment var thing and just try to load /etc/ihatemoney/ihatemoney.cfg but in this case I think we should create this directory by default on a production installation

No, in many cases, the user installing ihatemoney won't have the right to create something in /etc/.

0livd commented:
@JocelynDelalande @almet I added some comments a configuration section in the readme, what do you think ?

0livd commented:
It might also be a good idea to make the instance path configurable so we can change it in a dev environment

0livd commented:
@almet I added some tests for this new configuration process, there are still 2 todos related to the default config file though

almet commented on README.rst:
```diff
@@ -43,13 +33,43 @@ To deploy it, I'm using gunicorn and supervisord.
+In a production environment
+---------------------------
+
+Make a copy of budget/default_settings.py and name it ihatemoney.cfg.
```
nitpick: use double backquotes here to make file locations appear inside `pre` html tags. This applies to all file locations.

almet commented on README.rst:
```diff
@@ -43,13 +33,43 @@ To deploy it, I'm using gunicorn and supervisord.
+
+Make a copy of budget/default_settings.py and name it ihatemoney.cfg.
+Then adjust the settings to your needs and move this file to
+`/etc/ihatemoney/ihatemoney.cfg`.
```
nit: you should use double backquotes here instead of single ones.

almet commented on budget/run.py:
```diff
@@ -27,8 +27,18 @@ def pre_alembic_db():
+    # Load default settings first
+    # Then load the settings from the path set in IHATEMONEY_SETTINGS_FILE_PATH var
+    # If not set, default to /etc/ihatemoney/ihatemoney.cfg
+    # If the latter doesn't exist no errors are raised and the default settings are used
```
nit: no error *is* raised.

almet commented on budget/tests.py:
```diff
@@ -10,14 +10,18 @@
-
 from flask import session
 
+# Unset configuration file env var if previously set
```
Good catch ;)

almet commented:
This looks good to me!

If possible, we should address the TODO before merging to not add technical debt to the code, but all the rest looks great to me :-)

0livd commented:
I won't have time to push the last changes to the 2 pending PRs in the coming week so maybe you should merge them as is if you want to move forward in the meantime

almet commented:
Okay, thanks for letting us know, and thanks again for your work on the project!

almet commented:
I fixed the conflict in the README file.

0livd commented:
I rebased the PR and fixed the backquote nitpicks. For the TODOs, I need to create a fake ``ihatemoney.cfg`` so I'd like to do it after #196 is addressed

JocelynDelalande commented:
@0livd I don't understand how it depends on #196. Did you mean #193 ?

(Welcome back here:-))

0livd commented:
@JocelynDelalande These tests would add a new file :  ``budget/ihatemoney.cfg``, I would rather wait until we decide where to put the tests and put it directly in ``test/ihatemoney.cfg``. If I put it in the ``budget`` directory, users might think it is something actually used by the application.

0livd commented:
No more TODOs ! :)

almet commented:
Thanks! Seeing these changes, we probably should bump the major when releasing because it will break compatibility. Would that be okay for you? (cc @JocelynDelalande)
